qos-monitor/src/main/java/eu/arrowhead/core/qos/service/ping/monitor/impl/OrchestratedExternalPingMonitor.java

```diff
@@ -0,0 +1,270 @@
+			}
+		}
+
+		final int timeOut = calculateTimeOut();
+		final UUID measurementProcessId = requestExternalMeasurement(address);
+
+		final long startTime = System.currentTimeMillis();
+		final long meausermentExpiryTime = startTime + timeOut;
+
+		try {
+
+			return processor.processEvents(measurementProcessId, meausermentExpiryTime);
+
+		} catch (final Exception ex) {
+
+			logger.debug(ex.getMessage());
+		}
+
+		return null;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public void init() {
+		logger.debug("initPingMonitorProvider started...");
+
+		try {
+
+			final OrchestrationFormRequestDTO request = orchestrationRequestFactory.createExternalMonitorOrchestrationRequest();
+			final OrchestrationResponseDTO result = driver.queryOrchestrator(request);
+
+			cachedPingMonitorProvider = selectProvider(result);
+
+		} catch (final Exception ex) {
+			logger.warn("Exception in external ping monitor orchestration: " + ex);
+
+			cachedPingMonitorProvider = null;
+
+		}
+
+		threadPool.execute(eventCollector);
+
+		driver.unsubscribeFromPingMonitorEvents();
+		driver.subscribeToExternalPingMonitorEvents(getPingMonitorSystemRequestDTO());
+
+	}
+
+	//=================================================================================================
+	// assistant methods
+
+	//-------------------------------------------------------------------------------------------------
+	private UUID requestExternalMeasurement(final String address) {
+		logger.debug("requestExternalMeasurement started...");
+
+			try {
+				final IcmpPingRequestACK acknowledgedMeasurmentRequest = driver.requestExternalPingMonitorService(getPingMonitorProvidersServiceUri(), createIcmpPingRequest(address));
+				validateAcknowledgedMeasurmentRequest(acknowledgedMeasurmentRequest);
+
+				final UUID startedExternalMeasurementProcessId = acknowledgedMeasurmentRequest.getExternalMeasurementUuid();
+				logger.info("IcmpPingRequestACK received, with process id: " + startedExternalMeasurementProcessId);
+
+				return startedExternalMeasurementProcessId;
+
+			} catch (final Exception ex) {
+				logger.info(ex);
+
+				throw new ArrowheadException("External Ping Monitor is not available at: " + address );
+			}
+
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private void validateAcknowledgedMeasurmentRequest(final IcmpPingRequestACK acknowledgedMeasurmentRequest) {
+		logger.debug("validateAcknowledgedMeasurmentRequest started...");
+
+		try {
+			Assert.notNull(acknowledgedMeasurmentRequest, "IcmpPingRequestACK is null");
+			Assert.notNull(acknowledgedMeasurmentRequest.getAckOk(), "IcmpPingRequestACK.ackOk is null");
+			Assert.isTrue(acknowledgedMeasurmentRequest.getAckOk().equalsIgnoreCase("OK"), "IcmpPingRequestACK is null");
+		
+			
+		} catch (final Exception ex) {
+			logger.warn("External pingMonitorProvider replied invalid ack : " + ex);
+
```

Comment:
maybe you should nulled cachedPingMonitorProvider here

Reply:
fixed